nx/lib/nx/binary_backend/matrix.ex

```diff
@@ -174,44 +174,48 @@ defmodule Nx.BinaryBackend.Matrix do
   defp do_ts([], [], _idx, acc), do: acc
 
   def qr(input_data, input_type, input_shape, output_type, m, k, n, opts) do
-    {_, input_num_bits} = input_type
-
     mode = opts[:mode]
     eps = opts[:eps]
 
-    r_matrix =
-      if mode == :reduced do
-        # Since we want the first k rows of r, we can
-        # just slice the binary by taking the first
-        # n * k * output_type_num_bits bits from the binary.
-        # Trick for r = tensor[[0..(k - 1), 0..(n - 1)]]
-        slice_size = n * k * input_num_bits
-        <<r_bin::bitstring-size(slice_size), _::bitstring>> = input_data
-        binary_to_matrix(r_bin, input_type, {k, n})
+    {q_matrix, r_matrix} =
+      input_data
+      |> binary_to_matrix(input_type, input_shape)
+      |> qr_decomposition(m, n, eps)
+
+    {q_matrix, r_matrix} =
+      if mode == :reduced and m - k > 0 do
+        # Remove unnecessary columns (rows) from the matrix Q (R)
+        q_matrix_t = transpose_matrix(q_matrix)
+
+        {q_matrix_t, r_matrix} =
+          for _ <- 1..(m - k), reduce: {q_matrix_t, r_matrix} do
+            {q_t, r} ->
+              {List.delete_at(q_t, -1), List.delete_at(r, -1)}
```

Comment:
I think we need to remove more items here:
```suggestion
              {Enum.drop(q_t, k - m), Enum.drop(r, k - m)}
```

Not sure if we need a +1 there somewhere.
We also need a unit test with a matrix where m - n > 1 to show that this fix is needed

Reply:
Oh, nevermind, the for-comprehension deals with this!
We can substitute it with the suggested Enum.drop calls :)

Reply:
iex(1)> Enum.drop([1,2,3,4,5], -2)
[1, 2, 3]